source/postprocess/visualization.cc

```diff
@@ -127,6 +127,82 @@ namespace aspect
           }
       };
 
+
+      template <int dim>
+      class BaseAdjointVariablePostprocessor: public DataPostprocessor< dim >, public SimulatorAccess<dim>
+      {
+        public:
+          virtual
+          void
+          evaluate_vector_field(const DataPostprocessorInputs::Vector<dim> &input_data,
+                                std::vector<Vector<double> > &computed_quantities) const
+          {
+            const double velocity_scaling_factor =
+              this->convert_output_to_years() ? year_in_seconds : 1.0;
+            const unsigned int n_q_points = input_data.solution_values.size();
+            for (unsigned int q=0; q<n_q_points; ++q)
+              for (unsigned int i=0; i<computed_quantities[q].size(); ++i)
+                {
+                  // scale velocities and fluid velocities by year_in_seconds if needed
+                  if (this->introspection().component_masks.velocities[i] ||
+                      (this->include_melt_transport()
+                       && this->introspection().variable("fluid velocity").component_mask[i]))
+                    computed_quantities[q][i]=input_data.solution_values[q][i] * velocity_scaling_factor;
+                  else
+                    computed_quantities[q][i]=input_data.solution_values[q][i];
+                }
+          }
+
+          virtual std::vector<std::string> get_names () const
+          {
```

Comment:
I implemented this and it does work, but it then also adds two additional variables to the visual output, which are the adjoint compositional fields. Those don't really have any meaning and are just 0. If I keep the code here the same but pull from BaseVariablePostprocessor in the other functions I of course get an inconsistent number of variables (4 from the get_names and 6 from the others). Can you think of a fix for this?

Reply:
Hm, I see. The adjoint solution vector you store has these two additional compositional vector components, but what you are trying to do here is to just not output them at all. (In fact, I think you're dropping *all* compositional fields, right?) I had not noticed that. I guess that your previous approach makes sense in that case since you're just not using the adjoint compositional fields at all and never will because you're only solving the adjoint *Stokes* equations.

So go with your original approach, then. Sorry for the churn :-(